### PR TITLE
feat: Implement Dub.co for conversion tracking

### PR DESCRIPTION
**Pull request by steven-tey**
## What does this PR do?

This PR implements [Dub Conversions](https://dub.co/help/article/dub-conversions) to  measure how many new users (and upgrades) each existing Cal.com user brings in – also known as the K-factor or "viral coefficient formula".

### TODOs

- [x] Install the [`@dub/analytics` client-side JS SDK](https://dub.co/docs/conversions/quickstart#step-1-install-the-dub-analytics-client-side-sdk) for detecting the `dclid` query parameter and storing it as a first-party cookie, which will be used to attribute subsequent conversion events to the original link.
- [x] Install the [`dub` server-side SDK](https://dub.co/docs/conversions/quickstart#step-2-install-the-dub-server-side-sdk) for the following (hosted Cal version only):
	- Programmatically create `refer.cal.com/{username}` links when new users sign up
	- Send `lead` conversion events for new signups if `dclid` cookie is present – note that this uses [`cookies()`](https://nextjs.org/docs/app/api-reference/functions/cookies)  which is contingent on it being invoked in a Route Handler in the App Router.
- [x] Send a conversion event when a new user signs up (refer [NextAuth.js integration docs](https://dub.co/docs/conversions/leads/next-auth))
- [x] Create unique `refer.cal.com/{username}` short links on demand

### How to test.
1) First join the dub.co workspace
2) create yourself an API key 
3) Ensure HOSTED_CAL_FEATURES and IS_CALCOM constants are enabled
4) Signup a new user and copy referal link

### Viewing Conversions

Once this is set up, you should be able to see your conversions as a funnel chart view visualizing the conversion & dropoff rates across the different steps in the conversion funnel (clicks → leads → sales):

image

### Future enhancement

This can be extended to run a white-labeled referral program on Cal.com using Dub's [Analytics API](https://dub.co/docs/api-reference/endpoint/retrieve-analytics).

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

File: packages/features/auth/lib/next-auth-options.ts
Comment (sean-brydon) [inline packages/features/auth/lib/next-auth-options.ts:945]: I wonder how we can get this username stuff working with orgs. @hariombalhara  

For ref @steven-tey usernames arent unique anymore 

e.g. - cal.com/sean & i.cal.com/sean are different users.

Comment (hariombalhara) [reply packages/features/auth/lib/next-auth-options.ts:945]: We have similar handling at https://github.com/calcom/cal.com/blob/de46b1f1221f42e5a8399a00b698494406c15295/packages/features/auth/lib/next-auth-options.ts#L645 Maybe that can help.

Comment (steven-tey) [reply packages/features/auth/lib/next-auth-options.ts:945]: @sean-brydon maybe we can limit this to just `cal.com/{username}`? 

Also cc: @PeerRich who mentioned that we should use `refer.cal.com/{username}` instead of `go.cal.com/r/{username}` for the short links.

File: packages/features/shell/Shell.tsx
Comment (PeerRich) [inline packages/features/shell/Shell.tsx:956]: we should only show this on cal.com use IS_CALCOM from CONSTANTS

Comment (steven-tey) [reply packages/features/shell/Shell.tsx]: Fixed: https://github.com/calcom/cal.com/pull/16165/commits/2c3814d3804766831753754c67b522081a0bd218

File: packages/features/auth/lib/next-auth-options.ts
Comment (sean-brydon) [reply packages/features/auth/lib/next-auth-options.ts]: I think this is a fair shout - i think a lot of people in orgs will probably have the "powered by" bit disabled. From what i can see thats the main "referal" point. 

Nice job so far @steven-tey <3

File: apps/web/lib/app-providers-app-dir.tsx
Comment (zomars) [inline apps/web/lib/app-providers-app-dir.tsx:282]: This env variable is not available client-side. Are we ok with this?

File: apps/web/lib/app-providers.tsx
Comment (zomars) [inline apps/web/lib/app-providers.tsx:308]: Same here

**Review by zomars — COMMENTED** [review]
Looking good so far. Left some comments. Still need to test before merging.

File: apps/web/lib/app-providers-app-dir.tsx
Comment (steven-tey) [reply apps/web/lib/app-providers-app-dir.tsx]: Good catch – I'll get this updated 👍

Comment (steven-tey) [reply apps/web/lib/app-providers-app-dir.tsx]: Ended up hard-coding the domain but limiting this to only show when `IS_CALCOM` – LMK if that's okay!

File: apps/web/lib/app-providers-app-dir.tsx
Comment (sean-brydon) [inline apps/web/lib/app-providers-app-dir.tsx:286]: I think this check should also check for IS_CALCOM too right?

File: apps/web/lib/app-providers.tsx
Comment (sean-brydon) [inline apps/web/lib/app-providers.tsx:312]: Potentially same here

**Review by sean-brydon — COMMENTED** [review]
CleanShot 2024-08-19 at 11 13 54
Woo! it works! 

Great job man! I wasnt able to test the stripe "sales" side of this as i can't install the integration into dub. 

@zomars @keithwillcode @emrysal one of you guys may need to help steven with backdating links for users. Left two comments nothing blocking i'd say. I'll wait for zomars to test like he mentioned in his review 🙏

**Review by zomars — DISMISSED** [review]
Loving the new user workaround. We have some type checks failures before being able to merge tho. 🙏🏽 

image

File: apps/web/lib/app-providers-app-dir.tsx
Comment (steven-tey) [reply apps/web/lib/app-providers-app-dir.tsx:286]: I believe we are checking for `IS_CALCOM` in the conditional above, or am I misunderstanding something here? 👀

File: apps/web/lib/app-providers.tsx
Comment (steven-tey) [reply apps/web/lib/app-providers.tsx:312]: Same comment as above 🙏

File: apps/web/lib/app-providers-app-dir.tsx
Comment (sean-brydon) [reply apps/web/lib/app-providers-app-dir.tsx:286]: I think i miss understood what the domain option here is doing - makes sense now. Sorry :D

File: packages/features/auth/lib/next-auth-options.ts
Comment (sean-brydon) [inline packages/features/auth/lib/next-auth-options.ts:945]: @steven-tey What if we do the opposite here for creating links for existing users. 

Check if theyre not a new user - create a link ect. We should probably add a field to our user object to store referalLinks too. Lmk what you think

Comment (steven-tey) [reply packages/features/auth/lib/next-auth-options.ts:945]: @sean-brydon that's a great idea – we're thinking of doing something similar on Dub too, where we let users "request a referral link".
- When they do that for the first time, use `dub.links.create` to create a new link and store the linkId under `referralLinkId` field
- All subsequent requests will just copy the link as per their cal.com username (`refer.cal.com/:username`)

We still recommend [setting the `externalId` field when creating a link](https://dub.co/docs/quickstart/nextjs#3a-create-link-with-externalid) though since it gives you the flexibility to easily update the link later on using the `userId` instead of the `referralLinkId`.

Comment (steven-tey) [reply packages/features/auth/lib/next-auth-options.ts:945]: But with this approach, some of the users' booking pages will not have a valid `refer.cal.com/:username` link, is that okay? cc: @PeerRich 

In that case, it'll default back to the root `refer.cal.com` link, which still leads back to cal.com so it should be fine I think!

Comment (PeerRich) [reply packages/features/auth/lib/next-auth-options.ts:945]: sure. a fallback is nice, but would they not get a link the moment they log in and not have on?

!hasDubLink && createDubLink()

**Comment by PeerRich** [comment]
lets try and get this merged friday or monday, and then have a tagged release tuesday cc @zomars @keithwillcode 🙏

**Comment by PeerRich** [comment]
new PR merged for website

**Comment by PeerRich** [comment]
sorry for the delay

aiming at tuesday

**Comment by hbjORbj** [comment]
@steven-tey Hey Steven, the type-check CI job is failing because `WEBSITE_URL` is missing in import statements in `/app/web/pages/signup.tsx`. Can you add it? It can be imported from `@calcom/lib/constants`.

File: apps/web/lib/app-providers.tsx
Comment (hbjORbj) [inline apps/web/lib/app-providers.tsx:308]: Can you make the same change to `app-providers-app-dir.tsx`?

**Review by hbjORbj — DISMISSED** [review]

File: apps/web/lib/app-providers.tsx
Comment (steven-tey) [reply apps/web/lib/app-providers.tsx:308]: Update: I'll limit `DubAnlaytics` to just `/app/web/pages/signup.tsx` since that's the only page we need this script on – this should simplify things quite a bit!

**Review by sean-brydon — COMMENTED** [review]
Tested and LGTM.

 I'd be happy to merge this not sure if we want to get any more eyes here before we do @PeerRich @keithwillcode. 
 @steven-tey can you try restore the yarn.lock file. I tried but for some reason it wont change for me :')

**Comment by steven-tey** [comment]
@sean-brydon thanks for your help! I just fixed the `yarn.lock` file!

**Review by PeerRich — DISMISSED** [review]
approving on behalf of @sean-brydon

**Comment by zomars** [comment]
Checking for type errors

**Review by zomars — CHANGES_REQUESTED** [review]

**Review by zomars — APPROVED** [review]
All gucci now

**Pull request merged**
